plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/blobstore/S3BlobStoreContainerTests.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cloud.aws.blobstore;
+
+import com.amazonaws.Protocol;
+import com.amazonaws.services.s3.AmazonS3;
+import org.elasticsearch.cloud.aws.InternalAwsS3Service;
+import org.elasticsearch.common.blobstore.BlobStore;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.repositories.ESBlobStoreContainerTestCase;
+
+import java.io.IOException;
+import java.util.Locale;
+
+public class S3BlobStoreContainerTests extends ESBlobStoreContainerTestCase {
+    protected BlobStore newBlobStore() throws IOException {
+        int maxRetries = 5;
+        String region = null;
+        Settings settings = Settings.EMPTY;
+
+        InternalAwsS3Service service = new InternalAwsS3Service(settings);
+        AmazonS3 client = service.client(null, Protocol.HTTPS, region, null, null, maxRetries);
+        String bucket = randomAsciiOfLength(randomIntBetween(1, 10)).toLowerCase(Locale.ROOT);
+
+        return new S3BlobStore(settings, client, bucket, region, false,
+            new ByteSizeValue(100), maxRetries, "public-read-write", "standard");
+    }
```

Comment:
@gfyoung I would suggest simply creating a mock implementation of `AmazonS3` client for now.  This will involve lots of methods (see `AmazonS3Wrapper` which is just a wrapper for a real client, it has all the methods).  The good thing is, you will only have to implement a very select few for now.  I would suggest for the mock client, instead of hitting S3, we should just have an internal `Map` which "stores" the blob name -> blob data.  On this, you can fake out operations such as creating objects, deleting objects, etc. very easily as you're just working with a map.  See https://github.com/elastic/elasticsearch/blob/master/plugins/repository-azure/src/test/java/org/elasticsearch/cloud/azure/storage/AzureStorageServiceMock.java for inspiration.

Does that make sense?